File: ch10-sorting-and-searching/sort-algos/quickSort.js
function swap (arr, i, j) {
  if (i < j) {
    const temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
  }
}

function partition (arr, low, high) {
  let i = low - 1, j = low, pivot = arr[high]
  for (;j < high; j++) {
    if (arr[j] <= pivot) {
      i++
      swap(arr, i, j)
    }
  }
  i++
  swap(arr, i, high)
  return i
}

function quickSort (arr, low = 0, high = arr.length - 1) {
  if (low === 0 && high === 0) return arr

  if (low < high) {
    let piv = partition(arr, low, high)
    quickSort(arr, low, piv - 1)
    quickSort(arr, piv + 1, high)
  }

  return arr
}

// console.log(quickSort([1,3,2,7,5,4,5,3,4,1,6,8,3,2,2,4,1,6,7,2]))

File: implementations/algo/quick-sort.py
def partition(arr, lo, hi):
    pivot = arr[lo]
    i, j = lo + 1, hi

    while i <= j:
        while i <= j and arr[i] <= pivot:
            i += 1
        while i <= j and arr[j] >= pivot:
            j -= 1
        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

    arr[lo] = arr[j]
    arr[j] = pivot
    return j

def quick_sort(arr, lo, hi):
    if hi > lo:
        pivot_idx = partition(arr, lo, hi)
        quick_sort(array, lo, pivot_idx - 1)
        quick_sort(array, pivot_idx + 1, hi)

def quick_sort_main(arr):
    quick_sort(arr, 0, len(arr) - 1)
